### ISSUE TITLE
As a reader, I want to discover related articles to a topic I am reading about so that I can learn more about similar topics that I might otherwise miss.

### ISSUE THREAD

**Issue by jwflory** [unicef/inventory-hugo-theme#123]
### Summary

This is an exploratory issue. Are there ways we could better highlight what content is published in an article? Using tags is one way, and also making tagged content more discoverable could help.

~~There is also another potential here, like using a third-party library or tool to analyze the text and create a machine-generated summary. The priority of this user story is secondary, so we can look into this one as time allows. It could also be a good one for a new contributor to work on.~~ _Ruled out based on [these findings](https://github.com/unicef/inventory-hugo-theme/issues/123#issuecomment-1173052935)_.

See this section in the Hugo docs: [_Example: List Tags in a Single Page Template_](https://gohugo.io/templates/taxonomy-templates/#example-list-tags-in-a-single-page-template)

### Priority

secondary

### Category

front-end

### Type

functional

**Comment by BeeBombshell** [comment]
To summarize the available articles would be an excellent idea to enhance the readability and relevance of the user search. I came across the following npm library that summarizes text using an algorithm explained in the link below. I hope this helps! 😊 
- [Node-Summary NPM Package](https://www.npmjs.com/package/node-summary)

**Comment by jwflory** [comment]
@BeeBombshell This is super helpful, thanks for the share! 🙏🏻

**Comment by BeeBombshell** [comment]
Hi @jwflory, if this issue is open to contributions, should I give it a try? 😊

**Comment by jwflory** [comment]
Hey Bhavya, yes, you can take on this one! It's in our backlog so this is a good one to work on at your own pace. Do you need any information from me in order to get started?

Jun 22, 2022 13:56:42 Bhavya Verma ***@***.***>:

**Comment by BeeBombshell** [comment]
Thank you @jwflory. I will go through the project once to figure out how to proceed with the implementation. I'll discuss the same with you, along the way. If I have any queries or need any help, I'll reach out then. 😊

**Comment by jwflory** [comment]
@BeeBombshell This is neat, I like the direction you are going in.

> Kindly share your views about the idea. I was considering to create a function that could automatically generate keywords for the provided articles.

Before going too far with creating the function and integrating it into the site, I am curious what the user-facing interface for this would look like. How would we surface the tags in a way that help the reader quickly assess whether an article is relevant or not for them?

I am open to your creative ideas on what this could look like. One possible way is surfacing the tags in the index page of a category, and allowing the user to view articles sharing the same tags. Hugo includes [list page templates](https://gohugo.io/templates/lists) which would facilitate this nicely. This solution would need extra thinking on how the keywords would be extracted into the list page templates though.

There could also be other clever ways of doing this that I am not seeing either.

> Also, if you could provide a link for the sample articles in the project, it would be of great help! blush

Do you mean the example site, like [unicef.github.io/inventory-hugo-theme/](https://unicef.github.io/inventory-hugo-theme/)? It is worth noting that the example site will get a rehaul in the coming weeks (#89), so it may be more useful to try this out with the [Open Source Inventory](https://unicef.github.io/inventory/) site instead. This could also be very useful for us as we prepare to redesign the Open Source Inventory site around the DPG Standard too (#126).

**Comment by BeeBombshell** [comment]
@jwflory I like your idea of including list page templates to categorize the articles as per the tags. But yes, extracting the keywords for the user to search would be a tedious task. 

In addition, I was going through Hugo's documentation on the use of [Taxonomies](https://gohugo.io/content-management/taxonomies/). I think Hugo includes support for customized taxonomies, which could help in grouping the content as desired.

> I am open to your creative ideas on what this could look like. One possible way is surfacing the tags in the index page of a category, and allowing the user to view articles sharing the same tags.

For displaying these tags, one approach could be organizing the list of the articles in a `card` container, having the title on the top, followed by relevant tags and a few lines of description of the content. Given below is an example of what I'm imagining the component to look like. But I'm afraid that would lead to a lot of refactoring of the existing template.

![image](https://user-images.githubusercontent.com/53828745/175830387-5f66333e-2fbb-47d3-975e-02a3b251a8a0.png)

What are your thoughts about the same? 😄

**Comment by jwflory** [comment]
@BeeBombshell Thanks for sharing this example. I like your thinking, but I am wondering if we are biting off too much to chew. At least, for this first step.

Perhaps we could backtrack some and take a look at the Node-Summary package you linked earlier in the issue. If we started with the Node-Summary package, that would also give us a grip on how to bundle Node dependencies into the project, which currently does not utilize a Node runtime. I think the Node-Summary package meets the original need, which is summarizing an article into a condensed summary that is most relevant to the reader.

Could you try first by running the Node-Summary package on some of the [Open Source Inventory](https://unicef.github.io/inventory/) articles, perhaps any of these articles since they come up a lot:

* [Trademark strategies](https://unicef.github.io/inventory/legal-policy/trademark/)
* [Open Hardware case studies](https://unicef.github.io/inventory/hardware/case-studies/)
* [Project charter Mission](https://unicef.github.io/inventory/missions/charters/)

Knowing how these pages are summarized should be a good indicator of whether we proceed with this route or try something else. Any solution we devise here will venture into new territory by bundling in a Node dependency to the continuous integration pipeline.

**Comment by BeeBombshell** [comment]
Hi @jwflory, apologies for the late reply. 😅 

I tried running the suggested articles on the [Node Summary Package](https://www.npmjs.com/package/node-summary). The results are as follows:

- [Trademark strategies](https://unicef.github.io/inventory/legal-policy/trademark/)
![image](https://user-images.githubusercontent.com/53828745/177034558-ff4fb915-5868-491b-88cb-763fc5ab2a92.png)

- [Open Hardware case studies](https://unicef.github.io/inventory/hardware/case-studies/)
![image](https://user-images.githubusercontent.com/53828745/177034562-b3d11d95-0c57-456c-9ea0-588c62deff17.png)

- [Project charter Mission](https://unicef.github.io/inventory/missions/charters/)
![image](https://user-images.githubusercontent.com/53828745/177034572-0956c81c-894a-4de9-82d0-48e238f854ab.png)

The above results were generated by passing the URL for the provided articles. I also tried running the program by passing a .txt file for the same article. However, the outcome was quite irrelevant and lengthy.

I also tried a different package for summarization - [URL Summarizer](https://www.npmjs.com/package/url-summarizer).
The results were very similar to the previous package. The above two packages seem to be utilizing a similar kind of algorithm for summarizing the articles.

Finally, I found the [Text Summarization Package](https://www.npmjs.com/package/text-summarization) which produced a better outcome as compared to the previous trials. Following are the generated results for each of the three articles:

- [Trademark strategies](https://unicef.github.io/inventory/legal-policy/trademark/)
![image](https://user-images.githubusercontent.com/53828745/177035050-1b51ec2f-f1e1-4c6a-a6e6-182415c4d44e.png)

- [Open Hardware case studies](https://unicef.github.io/inventory/hardware/case-studies/)
![image](https://user-images.githubusercontent.com/53828745/177035056-b4bc6417-cbd1-4b06-99c6-b16e9ed56a86.png)

- [Project charter Mission](https://unicef.github.io/inventory/missions/charters/)
![image](https://user-images.githubusercontent.com/53828745/177035061-3fbd70d6-7762-414d-bd9c-5b36a1360beb.png)

What do you think about this? 😊

**Comment by jwflory** [comment]
Hey @BeeBombshell, thanks for digging in and researching this one!

This deep dive you prepared has revealed flawed assumptions about relying on machine-generated summaries. These summaries are not useful on their own, and if we did use them, they would require manual work to pick the "best" or right one that makes sense. (Another challenging part here is that the content published here (and in the other UNICEF toolkits) will use more technical language that could be difficult to summarize.)

So, I am thinking the best approach here based on your findings is to rely on manual, human-generated efforts of using **tags** in an article and then creating a better Hugo layout to discover content that shares the tags. This user story likely needs to be reframed in order to capture the new scope. Let me do that quickly.

**Comment by jwflory** [comment]
@BeeBombshell I put [this documentation reference](https://gohugo.io/templates/taxonomy-templates/#example-list-tags-in-a-single-page-template) in the original comment, but I wanted to highlight it here in the discussion too as a reference. The [list templates](https://gohugo.io/templates/lists/) page is also helpful. What do you think about this approach? Do you have ideas on what a layout could look like for the list-style layouts?

**Comment by BeeBombshell** [comment]
> This deep dive you prepared has revealed flawed assumptions about relying on machine-generated summaries. These summaries are not useful on their own, and if we did use them, they would require manual work to pick the "best" or right one that makes sense.

Hi @jwflory! Yes, I agree, we can't utilize the machine-generated summaries as of now. Even if we do find a tool that generates meaningful summaries, we can't trust a third-party application unless it's highly popular, and well-maintained.

> What do you think about this approach? Do you have ideas on what a layout could look like for the list-style layouts?

I think using the list templates is a nice approach. We can let the author create appropriate tags manually which can be displayed on the article's page (below the heading). We can then try to implement list templates on these tags to sort and display the required articles.

**Comment by jwflory** [comment]
@BeeBombshell Cool! We are aligned here on the approach.

I think the next steps for this issue would be to create a design wireframe/mockup and then use that to implement the change within Hugo. Would you want to take your hand at putting together a design wireframe? This could be a simple sketch, or something more polished with tools like Figma or Penpot.

**Comment by BeeBombshell** [comment]
Hi @jwflory, I created a simple wireframe to display how the tags could be put forth on the article pages. I'm not so experienced at creating wireframes. Hope this gives an idea of what I have in mind.

![Inventory theme tags](https://user-images.githubusercontent.com/53828745/181880300-bc6a357d-22b7-44fd-af58-4138e3c5db6f.png)

Kindly take a look and share your suggestions. 😄

**Comment by jwflory** [comment]
@BeeBombshell This is great! My only suggestion is that they could be a little smaller. It is good to call attention to them, but they shouldn't distract too much from the main text.

You are ready to give it a go on implementing this! How familiar with you Hugo are you? Do you need any pointers on how to get started? Let me know if I can give any pointers. 😀

**Comment by BeeBombshell** [comment]
Hi @jwflory, sorry for the delayed response, got caught up with some work. I'm not so familiar with using Hugo. I'd appreciate it if you could provide some resources to get started. 😄

**Comment by jwflory** [comment]
@BeeBombshell No problem.

You will need to create a new [partial template](https://gohugo.io/templates/partials/) in the theme (`/layouts/partials/`) and embed that into the [default partial template](https://github.com/unicef/inventory-hugo-theme/blob/main/layouts/partials/default.html). This will be part of a [single page template](https://gohugo.io/templates/single-page-templates/), which is the template used for rendering articles and published content on the site. The necessary steps are below:

1. Create a new partial template for tags, e.g. `/layouts/partials/page-tags.html`
2. Create the HTML text and change any CSS as needed in `/assets/css/style.css`
3. Add `{{ partial "page-tags.html" }}` into the [default partial template](https://github.com/unicef/inventory-hugo-theme/blob/main/layouts/partials/default.html)

To implement this feature, you will need [Hugo functions](https://gohugo.io/functions/) in your template to first check whether the page has any tags defined, and if it does, loop through them and create the tags you designed above so they appear for as many tags as an article has. The [`range` function](https://gohugo.io/functions/range/) is likely a good place to start.

Does this make sense? Let me know if anything is unclear.

**Comment by BeeBombshell** [comment]
Thanks, @jwflory this makes so much sense! 😄 
I'll get on the implementation ASAP. Will keep you updated, and let you know in case of any queries.

**Comment by jwflory** [comment]
@BeeBombshell Awesome, thanks so much! To help us with project management, could you also give your best estimate on when you would complete this? We work in two-week sprint cycles, with the current sprint ending this Friday (19 August). So we could target these following dates for the initial Pull Request:

- 19 August
- 9 September
- 23 September
- 7 October

Ideally it would be completed within this time range, but we could allot for more time if required. Let me know which target date you prefer! 😄

**Comment by BeeBombshell** [comment]
Hi @jwflory, I have started working on the implementation part. To be on the safer side, I'd say that I'll have this issue wrapped up by the 9th of September. For now, I need your advice on whether the following tags template looks alright. I've created a few sample tags to demonstrate how they'll look. I'll get to the mapping part next. 😊 

<img width="960" alt="image" src="https://user-images.githubusercontent.com/53828745/185067591-7a9ff386-5027-4f31-a455-8524e1f3e861.png">
